component/azstorage/block_blob.go

```diff
@@ -653,6 +656,181 @@ func (bb *BlockBlob) WriteFromBuffer(name string, metadata map[string]string, da
+		return &common.BlockOffsetList{}, err
+	}
+	for _, block := range *&storageBlockList.CommittedBlocks {
+		blk := &common.Block{
+			Id:         block.Name,
+			StartIndex: int64(blockOffset),
+			EndIndex:   int64(blockOffset) + block.Size,
+			Size:       block.Size,
+		}
+		blockOffset += block.Size
+		blockList.BlockOffsetList = append(blockList.BlockOffsetList, blk)
+	}
+	if blockOffset == 0 {
+		blockList.SmallFile = true
+	}
+	return &blockList, nil
+}
+
+// create our definition of block
+func (bb *BlockBlob) createBlock(blockIdLength, startIndex, size int64) *common.Block {
+	newBlockId := base64.StdEncoding.EncodeToString(common.NewUUID(blockIdLength))
+	newBlock := &common.Block{
+		Id:         newBlockId,
+		StartIndex: startIndex,
+		EndIndex:   startIndex + size,
+		Size:       size,
+	}
+	return newBlock
+}
+
+func (bb *BlockBlob) createNewBlocks(modBlockList, blockList *common.BlockOffsetList, offset, length, blockIdLength int64) (bool, int64) {
+	prevIndex := blockList.BlockOffsetList[len(blockList.BlockOffsetList)-1].EndIndex
+	// BufferSize is the size of the buffer that will go beyond our current blob (appended)
+	var bufferSize int64
+	// appendOnly means there is no overlap at all with the blob - data only being appended
+	appendOnly := len(modBlockList.BlockOffsetList) == 0
+	for i := prevIndex; i < offset+length; i += bb.Config.blockSize {
+		// create a new block if we hit our block size
+		blkSize := int64(math.Min(float64(bb.Config.blockSize), float64((offset+length)-i)))
+		newBlock := bb.createBlock(blockIdLength, i, blkSize)
+		modBlockList.BlockOffsetList = append(modBlockList.BlockOffsetList, newBlock)
+		blockList.BlockOffsetList = append(blockList.BlockOffsetList, newBlock)
+		// reset the counter since it will help us to determine if there is leftovers at the end
+		bufferSize += blkSize
+	}
+	return appendOnly, bufferSize
+}
+
+// Write : write data at given offset to a blob
+func (bb *BlockBlob) Write(name string, offset, length int64, data []byte, fileOffsets, modFileOffsets *common.BlockOffsetList) error {
+	defer log.TimeTrack(time.Now(), "BlockBlob::Write", name)
+	log.Trace("BlockBlob::Write : name %s offset %v", name, offset)
+	// tracks the case where our offset is great than our current file size (appending only - not modifying pre-existing data)
+	appendOnly := false
+	var dataBuffer *[]byte
+
+	// when the file offset mapping is cached we don't need to make a get block list call
+	if fileOffsets != nil && !fileOffsets.Cached {
+		var err error
+		fileOffsets, err = bb.GetFileBlockOffsets(name)
+		if err != nil {
+			return err
+		}
+	}
+
+	// case 1: file consists of no blocks (small file)
+	if fileOffsets != nil && fileOffsets.SmallFile {
+		// get all the data
+		oldData, _ := bb.ReadBuffer(name, 0, 0)
+		// update the data with the new data
+		// if we're only overwriting existing data
+		if int64(len(oldData)) >= offset+length {
```

Comment:
we can potentially reduce this if else if we knew the old data size before hand (file size)